File: Source/SiteCommon/Web/services/datastore.ts
import {Dictionary} from "../base/Dictionary"
import MainService from "./mainservice";


export class DataStore {
    PublicDataStore: Dictionary<Dictionary<any>>;
    PrivateDataStore: Dictionary<Dictionary<any>>;
    private MS: MainService;

    CurrentRoutePage: string ='';
    DeploymentIndex: string = '';

    constructor(MainService) {
        this.MS = MainService;
        this.PrivateDataStore = new Dictionary<Dictionary<any>>();
        this.PublicDataStore = new Dictionary<Dictionary<any>>();
        this.loadDataStores();
    }

    toJSON() {
        var toConvert:any = {};
        toConvert.PublicDataStore = this.PublicDataStore;
        toConvert.PrivateDataStore = this.PrivateDataStore;
        toConvert.CurrentRoutePage = this.CurrentRoutePage;
        toConvert.DeploymentIndex = this.DeploymentIndex;
        return toConvert;
    }

    loadDataStoreFromJson(value: any) {
        if (!value) {
            return;
        }

        let privateStore = value.PrivateDataStore;
        let publicStore = value.PublicDataStore;

        if (privateStore) {
            this.PrivateDataStore = new Dictionary<Dictionary<any>>();
            for (let route in privateStore) {
                let valueToAdd: Dictionary<any> = new Dictionary<any>();
                for (let key in privateStore[route]) {
                    valueToAdd.add(key, privateStore[route][key]);
                }
                this.PrivateDataStore.add(route, valueToAdd);
            }
        }
       
        if (publicStore) {
            this.PublicDataStore = new Dictionary<Dictionary<any>>();
            for (let route in publicStore) {
                let valueToAdd: Dictionary<any> = new Dictionary<any>();
                for (let key in publicStore[route]) {
                    valueToAdd.add(key, publicStore[route][key]);
                }
                this.PublicDataStore.add(route, valueToAdd);
            }
        }
    }
   

    public currentRoute(): string {
        return this.CurrentRoutePage + "-" + this.DeploymentIndex;
    }

    /// This method will be used on startup from the main service
    private loadDataStores() {
        let datastore:any = this.MS.UtilityService.GetItem(this.MS.NavigationService.appName + " datastore");
        if (!datastore) {
            this.PublicDataStore = new Dictionary<Dictionary<any>>();
            this.PrivateDataStore = new Dictionary<Dictionary<any>>();
        } else {
            this.loadDataStoreFromJson(datastore);
        }
    }


    private cacheDataStores() {
        this.MS.UtilityService.SaveItem(this.MS.NavigationService.appName + " datastore", this);
    }


    public routeExists(route: string, dataStoreType: DataStoreType = DataStoreType.Any): boolean {
        var found: boolean = false;

        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
            if (this.PrivateDataStore.containsKey(route)) {
                found = true;
            }
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            if (this.PublicDataStore.containsKey(route)) {
                found = true;
            }
        }
        return found;
    }

    public keyExists(key: string, dataStoreType: DataStoreType = DataStoreType.Any): boolean {
        return this.getValueAndRoutesFromDataStore(dataStoreType, key).length > 0;
    }

    public routeAndKeyExists(route: string, key:string, dataStoreType: DataStoreType = DataStoreType.Any): boolean {
        var found: boolean = false;

        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
            if (this.PrivateDataStore.containsKey(route)) {
                if (this.PrivateDataStore.get(route).containsKey(key)) {
                    found = true;
                }
            }
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            if (this.PublicDataStore.containsKey(route)) {
                if (this.PublicDataStore.get(route).containsKey(key)) {
                    found = true;
                }
            }
        }
        return found;
    }

    public addToDataStoreWithCustomRoute(route:string, key: string, value: any, dataStoreType: DataStoreType) {
        this.updateValue(dataStoreType, this.currentRoute(), key, value);
        this.cacheDataStores();
    }

    public addToDataStore(key:string, value: any, dataStoreType: DataStoreType) {
        this.updateValue(dataStoreType, this.currentRoute(), key, value);
        this.cacheDataStores();
    }

    public addObjectToDataStore(value: any, dataStoreType: DataStoreType)
    {
        for (let propertyName in value) {
            this.updateValue(dataStoreType, this.currentRoute(), propertyName, value[propertyName]);
        }
        this.cacheDataStores();
    }

    public addObjectToDataStoreWithCustomRoute(route:string, value: any, dataStoreType: DataStoreType) {
        for (let propertyName in value) {
            this.updateValue(dataStoreType, route, propertyName, value[propertyName]);
        }
        this.cacheDataStores();
    }

    public getJson(key: string, dataStoreType: DataStoreType = DataStoreType.Any): any {
        return this.getFirstValueFromDataStore(key, dataStoreType);
    }

    public getJsonWithRoute(route:string, key: string, dataStoreType: DataStoreType = DataStoreType.Any): any {
        return this.getValueWithRouteAndKey(dataStoreType, route, key);
    }

    public getValue(key: string, dataStoreType: DataStoreType = DataStoreType.Any): string {
        return this.getFirstValueFromDataStore(key, dataStoreType).toString();
    }

    public getValueWithRoute(route: string, key: string, dataStoreType: DataStoreType = DataStoreType.Any): string {
        return this.getValueWithRouteAndKey(dataStoreType, route, key).toString();
    }

    public getAllJson(key: string, dataStoreType: DataStoreType = DataStoreType.Any):any[] {
        return this.getAllValueFromDataStore(key, dataStoreType);
    }

    public getAllValues(key: string, dataStoreType: DataStoreType = DataStoreType.Any):string[] {
        return this.getAllValueFromDataStore(key, dataStoreType).map(p => p.toString());
    }

    public getAllDataStoreItems(key: string, dataStoreType: DataStoreType = DataStoreType.Any):DataStoreItem[] {
        return this.getValueAndRoutesFromDataStore(dataStoreType, key);
    }

    private getFirstValueFromDataStore(key: string, dataStoreType: DataStoreType = DataStoreType.Any):any {
        var values: DataStoreItem[];

        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
            values = DataStore.getValueAndRoutesFromDataStore(this.PrivateDataStore, key, DataStoreType.Private);
            if (values.length >0 ) {
                return values[0].value;
            }
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            values = DataStore.getValueAndRoutesFromDataStore(this.PublicDataStore, key, DataStoreType.Private);
            if (values.length > 0 ) {
                return values[0].value;
            }
        }

        return null;
    }

    private getAllValueFromDataStore(key: string, dataStoreType: DataStoreType = DataStoreType.Any):any[] {
        var items: DataStoreItem[]  = this.getValueAndRoutesFromDataStore(dataStoreType, key);

        return items.map((value, index, array) => { return value.value });
    }

    private getValueAndRoutesFromDataStore(dataStoreType: DataStoreType, key: string): DataStoreItem[] {
        var valuesToReturn: DataStoreItem[] = new Array<DataStoreItem>();

        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {

            valuesToReturn = valuesToReturn.concat(DataStore.getValueAndRoutesFromDataStore(this.PrivateDataStore, key, DataStoreType.Private));
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            valuesToReturn = valuesToReturn.concat(DataStore.getValueAndRoutesFromDataStore(this.PublicDataStore, key, DataStoreType.Public));
        }

        return valuesToReturn;
    }

    private getValueWithRouteAndKey(dataStoreType: DataStoreType, route: string, key: string) {
        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
            if (this.PrivateDataStore.containsKey(route) && this.PrivateDataStore.get(route).containsKey(key)) {
                return this.PrivateDataStore.get(route).get(key);
            }
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            if (this.PublicDataStore.containsKey(route) && this.PublicDataStore.get(route).containsKey(key)) {
                return this.PublicDataStore.get(route).get(key);
            }
        }

        return null;
    }

    private updateValue(dataStoreType: DataStoreType, route: string, key: string, value: any) {
        var foundInPrivate: boolean = false;
        var foundInPublic: boolean = false;

        if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
            foundInPrivate = DataStore.updateItemIntoDataStore(this.PrivateDataStore, route, key, value);
        }

        if (dataStoreType === DataStoreType.Public || dataStoreType === DataStoreType.Any) {
            foundInPublic = DataStore.updateItemIntoDataStore(this.PublicDataStore, route, key, value);
        }

        if (!foundInPublic && !foundInPrivate) {
            if (dataStoreType === DataStoreType.Private || dataStoreType === DataStoreType.Any) {
                DataStore.addModifyItemToDataStore(this.PrivateDataStore, route, key, value);
            }

            if (dataStoreType === DataStoreType.Public) {
                DataStore.addModifyItemToDataStore(this.PublicDataStore, route, key, value);
            }
        }
    }

    private static getValueAndRoutesFromDataStore(store: Dictionary<Dictionary<any>>,
        key: string,
        dataStoreType: DataStoreType): DataStoreItem[] {
        var itemsMatching = new Array<DataStoreItem>();

        for (var i = 0; i < store.length(); i++) {
            var item: [string, Dictionary<any>] = store.getItem(i);

            if (item["1"].containsKey(key)) {
                var itemToAdd: DataStoreItem = new DataStoreItem();
                itemToAdd.route = item["0"];
                itemToAdd.key = key;
                itemToAdd.value = item["1"].get(key);
                itemsMatching.push(itemToAdd);
            }
        }

        return itemsMatching;
    }

    private static updateItemIntoDataStore(store: Dictionary<Dictionary<any>>,
        route: string,
        key: string,
        value: any): boolean {
        var found: boolean = false;

        if (store.containsKey(route) && store.get(route).containsKey(key)) {
            found = true;
            if (value === null) {
                store.get(route).remove(key);
            }
            else {
                store.get(route).modify(key, value);
            }
        }

        return found;
    }

    private static addModifyItemToDataStore(store: Dictionary<Dictionary<any>>,
        route: string,
        key: string,
        value: any): void {

        if (!store.containsKey(route)) {
            store.add(route, new Dictionary<any>());
        }

        if (!store.get(route).containsKey(key)) {
            store.get(route).add(key, value);
        }

        store.get(route).modify(key, value);
    }
}

export class DataStoreItem {
    route: string;
    key: string;
    value: any;
}

export enum DataStoreType {
    Public,
    Private,
    Any
}

File: Source/Common/Microsoft.Deployment.Common/ActionModel/DataStore.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;

namespace Microsoft.Deployment.Common.ActionModel
{
    /// <summary>
    /// This class stores both public and private details for the application
    /// This class should resemble the typescript to ensure consistency
    /// </summary>
    public class DataStore
    {
        public string CurrentRoutePage { get; set; }
        public string DeploymentIndex { get; set; }

        public string CurrentRoute => CurrentRoutePage + "-" + DeploymentIndex;

        public Dictionary<string,Dictionary<string,JToken>> PublicDataStore { get; set; } = new Dictionary<string, Dictionary<string, JToken>>();

        public Dictionary<string, Dictionary<string, JToken>> PrivateDataStore { get; set; } = new Dictionary<string, Dictionary<string, JToken>>();

        // Removed for the time being to force usage of getfirst and getall
        //public JToken this[string route, string key]
        //{
        //    get
        //    {
        //        return this.GetValueWithRouteAndKey(DataStoreType.Any, route, key);
        //    }
        //    set
        //    {
        //        this.UpdateValue(DataStoreType.Public, route, key, value);
        //    }
        //}

        //public JToken this[DataStoreType type, string route, string key]
        //{
        //    get
        //    {
        //        return this.GetValueWithRouteAndKey(type, route, key);
        //    }
        //    set
        //    {
        //        this.UpdateValue(type, route, key, value);
        //    }
        //}

        //public IList<DataStoreItem> this[string key, DataStoreType dataStoreType = DataStoreType.Any]
        //{
        //    get
        //    {
        //        return this.GetValueAndRoutesFromDataStore(DataStoreType.Any, key);
        //    }
        //}

        //public IList<DataStoreItem> this[DataStoreType type, string key]
        //{
        //    get
        //    {
        //        return this.GetValueAndRoutesFromDataStore(type, key);
        //    }
        //}

        public bool RouteExists(string route, DataStoreType dataStoreType = DataStoreType.Any)
        {
            bool found = false;

            if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
            {
                if (PrivateDataStore.ContainsKey(route))
                {
                    found = true;
                }
            }

            if (dataStoreType == DataStoreType.Public || dataStoreType == DataStoreType.Any)
            {
                if (PublicDataStore.ContainsKey(route))
                {
                    found = true;
                }
            }

            return found;
        }

        public bool KeyExists(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetValueAndRoutesFromDataStore(dataStoreType, key).Any();
        }

        public bool RouteAndKeyExists(string route, string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            var listOfKeys = this.GetValueAndRoutesFromDataStore(dataStoreType, key).ToList();
            if (listOfKeys.Any())
            {
                return listOfKeys.Any(p => p.Route.Equals(route));
            }

            return false;
        }


        public void AddToDataStore(string route, string key, JToken value, DataStoreType dataStoreType = DataStoreType.Any)
        {
            this.UpdateValue(dataStoreType, route, key, value);
        }

        public void AddToDataStore(string key, JToken value, DataStoreType dataStoreType  = DataStoreType.Any)
        {
            this.UpdateValue(dataStoreType, this.CurrentRoute, key, value);
        }

        public void AddObjectDataStore(string route, JObject value, DataStoreType dataStoreType)
        {
            foreach (var val in value)
            {
                this.UpdateValue(dataStoreType, route, val.Key, val.Value);
            }
        }

        public void AddObjectDataStore(JObject value, DataStoreType dataStoreType)
        {
            foreach (var val in value)
            {
                this.UpdateValue(dataStoreType, this.CurrentRoute, val.Key, val.Value);
            }
        }

        public string GetValue(string route, string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetValueWithRouteAndKey(dataStoreType, route, key)?.ToString();
        }

        public JToken GetJson(string route, string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetValueWithRouteAndKey(dataStoreType, route, key);
        }

        public string GetValue(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetFirstValueFromDataStore(key, dataStoreType)?.ToString();
        }

        public JToken GetJson( string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetFirstValueFromDataStore(key, dataStoreType);
        }

        public DataStoreItem GetDataStoreItem(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetValueAndRoutesFromDataStore(dataStoreType, key).First();
        }

        public IList<string> GetAllValues(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetAllValueFromDataStore(key,dataStoreType).Select(p=>p?.ToString()).ToList();
        }

        public IList<JToken> GetAllJson(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetAllValueFromDataStore(key, dataStoreType);
        }

        public IList<DataStoreItem> GetAllDataStoreItems(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            return this.GetValueAndRoutesFromDataStore(dataStoreType,key);
        }

        private JToken GetFirstValueFromDataStore( string key, DataStoreType dataStoreType = DataStoreType.Any) 
        {
            if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
            {
                var values = GetValueAndRoutesFromDataStore(this.PrivateDataStore, key, DataStoreType.Private);
                if (values.Any())
                {
                    return values.First().Value;
                }
            }

            if (dataStoreType == DataStoreType.Public || dataStoreType == DataStoreType.Any)
            {
                var values = GetValueAndRoutesFromDataStore(this.PublicDataStore, key, DataStoreType.Private);
                if (values.Any())
                {
                    return values.First().Value;
                }
            }

            return null;
        }

        private IList<JToken> GetAllValueFromDataStore(string key, DataStoreType dataStoreType = DataStoreType.Any)
        {
            IList<DataStoreItem> items = GetValueAndRoutesFromDataStore(dataStoreType, key);
            return items.Select(p => p.Value).ToList();
        }

        private IList<DataStoreItem> GetValueAndRoutesFromDataStore(DataStoreType dataStoreType, string key)
        {
            List<DataStoreItem> valuesToReturn = new List<DataStoreItem>();

            if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
            {
                valuesToReturn.AddRange(GetValueAndRoutesFromDataStore(this.PrivateDataStore, key, DataStoreType.Private));
            }

            if (dataStoreType == DataStoreType.Public || dataStoreType == DataStoreType.Any)
            {
                valuesToReturn.AddRange(GetValueAndRoutesFromDataStore(this.PublicDataStore, key, DataStoreType.Public));
            }

            return valuesToReturn;
        }

        private JToken GetValueWithRouteAndKey(DataStoreType dataStoreType, string route, string key)
        {
            if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
            {
                if (PrivateDataStore.ContainsKey(route) && PrivateDataStore[route].ContainsKey(key))
                {
                    return PrivateDataStore[route][key];
                }
            }

            if (dataStoreType == DataStoreType.Public || dataStoreType == DataStoreType.Any)
            {
                if (PublicDataStore.ContainsKey(route) && PublicDataStore[route].ContainsKey(key))
                {
                    return PublicDataStore[route][key];
                }
            }

            return null;
        }

        public void UpdateValue(DataStoreType dataStoreType, string route, string key, JToken value)
        {
            bool foundInPrivate = false;
            bool foundInPublic = false;

            if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
            {
                foundInPrivate = UpdateItemIntoDataStore(this.PrivateDataStore, route, key, value);
            }

            if (dataStoreType == DataStoreType.Public || dataStoreType == DataStoreType.Any)
            {
                foundInPublic=  UpdateItemIntoDataStore(this.PublicDataStore, route, key, value);
            }

            if (!foundInPublic && !foundInPrivate)
            {
                if (dataStoreType == DataStoreType.Private || dataStoreType == DataStoreType.Any)
                {
                    AddModifyItemToDataStore(this.PrivateDataStore, route,key,value);
                }

                if (dataStoreType == DataStoreType.Public)
                {
                    AddModifyItemToDataStore(this.PublicDataStore, route, key, value);
                }
            }
        }

        private static List<DataStoreItem> GetValueAndRoutesFromDataStore(Dictionary<string, Dictionary<string, JToken>> store,
            string key, DataStoreType dataStoreType)
        {
            List<DataStoreItem> itemsMatching = new List<DataStoreItem>();

            foreach (var keyValuePair in store)
            {
                if (keyValuePair.Value.ContainsKey(key))
                {
                    itemsMatching.Add(new DataStoreItem()
                    {
                        Route = keyValuePair.Key,
                        Key = key,
                        Value = keyValuePair.Value[key],
                        DataStoreType = dataStoreType
                    });
                }
            }

            return itemsMatching;
        }

        private static bool UpdateItemIntoDataStore(Dictionary<string, Dictionary<string, JToken>> store,
            string route, string key, JToken value)
        {
            bool found = false;
            if (store.ContainsKey(route) && store[route].ContainsKey(key))
            {
                found = true;
                if (value == null)
                {
                    store[route].Remove(key);

                    // Bug - TODO fix this area
                    //if (!store[key].Any())
                    //{
                    //    store.Remove(route);
                    //}
                }
                else
                {
                    store[route][key] = value;
                }
            }

            return found;
        }

        private static void AddModifyItemToDataStore(Dictionary<string, Dictionary<string, JToken>> store,
            string route, string key, JToken value)
        {
            if (!store.ContainsKey(route))
            {
                store.Add(route, new Dictionary<string, JToken>());
            }

            if (!store[route].ContainsKey(key))
            {
                store[route].Add(key, value);
            }

            store[route][key] = value;
        }
    }
}
